bvm/ballerina-core/src/main/java/org/ballerinalang/model/util/XMLUtils.java

```diff
@@ -365,6 +388,54 @@ public static BValue convertToJSON(BXML xml, String attributePrefix, boolean pre
         return json;
     }
 
+    /**
+     * Compares if two xml values are equal.
+     *
+     * Equality is computed as follows
+     * - for XML elements: compares the canonicalized versions, including comments
+     * - for non-elements (standalone text, PI, comments): a string comparison
+     *
+     * @param xmlOne    the first XML value
+     * @param xmlTwo    the second XML value
+     * @return  true if the two are equal, false if not
+     */
+    public static boolean isEqual(BXML<?> xmlOne, BXML<?> xmlTwo) {
+        XMLNodeType xmlOneNodeType = xmlOne.getNodeType();
+        XMLNodeType xmlTwoNodeType = xmlTwo.getNodeType();
+
+        if (xmlOneNodeType == XMLNodeType.SEQUENCE && xmlTwoNodeType == XMLNodeType.SEQUENCE) {
```

Comment:
There can be a situation where one is a sequence of size 1 (singleton, a sequence with a single element type xml in it), and the other is an element type xml (singleton, but not a sequence). We need to treat both of them as singletons.